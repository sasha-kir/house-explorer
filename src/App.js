import React, { Component } from 'react';
import './App.css';

import Navigation from './components/navigation/Navigation';

class App extends Component {
  render() {
    return (
      <div className="App">
        <Navigation /> 
        {/* 
        <LandingBody />

        <SearchBar />
        <Map />
        <HouseInfoBlock />
        <PhotoBlock />

        <History />
        */}
      </div>
    );
  }
}

export default App;
