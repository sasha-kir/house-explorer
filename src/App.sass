@import config

.body
    display: grid
    grid-template-rows: 1fr 5fr 1fr
    height: 100vh
    background-color: $primaryColor