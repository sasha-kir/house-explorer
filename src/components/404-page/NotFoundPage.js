import React from 'react';
import '../placeholder-page/PlaceholderPage.sass';

import notFound from '../../images/not-found.png';

const NotFoundPage = () => {
    return (
        <div className="not-found-main-div">
            <p>page not found (._.)</p>
            <div className="not-found-image-wrapper">
                <img alt="404 message illustration" src={notFound} />
                <span className="not-found-image-credit">&copy; &nbsp;
                    <a href="https://icons8.com/ouch/illustration/marginalia-page-not-found" 
                        rel="noopener noreferrer" target="_blank">Icons 8</a>
            </span>
            </div>
        </div>
    );
}

export default NotFoundPage;