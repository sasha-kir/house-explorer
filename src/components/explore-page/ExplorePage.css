.explore-main-div {
    background-color: #ABC6FD;
    margin-top: 100px;
    display: grid;
}