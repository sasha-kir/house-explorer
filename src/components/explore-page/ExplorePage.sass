@import '../../_config.sass'

.explore-main-div
    @extend %page-template
    background-color: $explorePrimaryColor
    grid-template-columns: 2.5fr 3fr
    padding-bottom: 25px
    
.explore-content
    padding: 5px 0 15px 0
    display: grid
    grid-template-rows: 1fr 7fr
    height: $mainMapHeight