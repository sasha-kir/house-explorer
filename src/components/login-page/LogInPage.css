.login-main-div {
    height: 100vh;
    display: flex;
    justify-content: center;
    background-color: #ABC6FD;
     /* http://www.heropatterns.com/ */
    background-image: url("data:image/svg+xml,%3Csvg \
    width='42' height='44' viewBox='0 0 42 44' \
    xmlns='http://www.w3.org/2000/svg'%3E%3Cg \
    id='Page-1' fill='none' fill-rule='evenodd'%3E%3Cg \
    id='brick-wall' fill='%23f3f5e7' \
    fill-opacity='0.33'%3E%3Cpath d='M0 0h42v44H0V0zm1 \
    1h40v20H1V1zM0 23h20v20H0V23zm22 \
    0h20v20H22V23z'/%3E%3C/g%3E%3C/g%3E%3C/svg%3E");
}

.login-wrapper {
    background-color: #ABC6FD;
    align-self: center;
    height: 410px;
    width: 450px;
    display: grid;
    grid-template-rows: 1fr 2fr;
    box-shadow: 0 0 10px rgba( 255, 255, 255, 0.5 );
    /* border: 2px solid #383A49; */
}

/* .logo-image {
    justify-self: center;
    align-self: end;
    height: 100px;
    width: auto;
    padding-top: 30px;
} */

.login-form {
    display: grid;
    padding: 50px;
    row-gap: 10px;
    font-size: 30px;
    grid-template-rows: 1fr 1fr 1fr 1fr;
}

.form-header {
    font-size: 45px;
    font-variant: small-caps;
    margin: 0 0 20px 0;
    letter-spacing: .1em;
    text-align: center;
}

.form-element {
    background: transparent;
    padding-left: 10px;
    align-self: center;
    outline: 0;
    border-width: 0 0 2px;
    border-color: #383A49;
    max-height: 40px;
    min-height: 40px;
    font-size: 20px;
}

.form-element::placeholder {
    color: rgba(56, 58, 73, 0.3);
}

.form-element, 
.form-button {
    color: #383A49;
}

.form-button {
    background-color: transparent;
    font-variant: small-caps;
    letter-spacing: .1em;
    border: 3px solid #383A49;
    border-radius: 5px;
    justify-self: stretch;
    align-self: end;
    padding: 5px 10px;
    margin-top: 20px;
    font-size: 25px;
    max-height: 50px;
    min-height: 50px;
}

.form-button:hover {
    color: #F3F5E7;
    background-color: #383A49;
    transition: .25s ease-in;
}