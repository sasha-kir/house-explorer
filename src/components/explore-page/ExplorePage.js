import React, { Component } from 'react';
import { Helmet } from "react-helmet-async";
import { YMaps } from "react-yandex-maps";

import { cityList } from './cityList';
import './ExplorePage.sass';

import SearchBar from '../search-bar/SearchBar';
import HouseInfoBlock from '../house-info-block/HouseInfoBlock';

const YandexMap = React.lazy(() => 
    import ('../yandex-map/YandexMap')
);


class ExplorePage extends Component {
    constructor() {
        super();
        this.state = {
            inputValue: "",                             // input contents
            searchTerm: "",                             // what is searched on submit
            savedInput: "",                             // what user unitially typed, without suggestion
            searchSuggestions: [],                      // suggestions fetched from dadata
            locationInRussian: ["Москва", "Россия"],    // [city, country] in Russian
            locationInEnglish: ["Moscow", "RU-MOW"],    // [city, isoCode] in English
            startState: true,                           // display instructions or no
            addressNotFound: false,                     // if address was not found by dadata
            mapCoords: [0, 0],                          // coordinates to place pin and map center
            mapPinAddress: "Москва, Россия",            // full address at pin
            infoBlockAddress: ""                        // address in info block
        };
        this.dadataAPIKey = process.env.REACT_APP_DADATA_API_KEY;
        this.yandexAPIKey = process.env.REACT_APP_YANDEX_API_KEY;
    }

    getCityFromIP = (ip) => {
        const url = `https://suggestions.dadata.ru/suggestions/api/4_1/rs/iplocate/address?ip=${ip}`
        fetch(url, {
            method: 'GET',
            headers: {
                "Accept": "application/json",
                "Authorization": `Token ${this.dadataAPIKey}`
            }
        })
            .then(response => response.json())
            .then(data => {
                if (data.location) {
                    const { city, country, region_iso_code } = data.location.data;
                    this.setState({ 
                                    locationInRussian: [city, country],
                                    mapPinAddress: `${city}, ${country}`,
                                    locationInEnglish: ["", region_iso_code]
                                });
                }
            })
            .catch(console.log)
    }

    componentDidMount() {
        // get IP and starting mapCoords
        let userIP;
        fetch("http://ip-api.com/json")
            .then(result => result.json())
            .then(data => {
                const { lat, lon } = data;
                userIP = data.query;
                this.setState({ mapCoords: [lat, lon] });
            })
            .then(() => this.getCityFromIP(userIP))
            .catch(console.log)
    }

    handleSearchInput = (inputValue, translit = "") => {
        if (translit) {
            this.setState({ searchTerm: translit });
        } else {
            this.setState({ searchTerm: inputValue });
        };

        this.setState({ inputValue: inputValue });
    }

    clearSearchTerm = () => {
        this.setState({ searchTerm: "", inputValue: "" });
    }

    sendDadataRequest = (query, count) => {
        const url = "https://suggestions.dadata.ru/suggestions/api/4_1/rs/suggest/address";
        const { locationInRussian } = this.state;
        return fetch(url, {
                    method: 'POST',
                    headers: {
                        'Content-Type': 'application/json',
                        'Accept': 'application/json',
                        'Authorization': `Token ${this.dadataAPIKey}`
                    },
                    body: JSON.stringify({
                        'query': query,
                        'count': count,
                        'locations': [
                            { 
                                "city": locationInRussian[0],
                                "country": locationInRussian[1]
                            }
                        ],
                        'restrict_value': true
                    })
            })
                .then(response => response.json())
                .then(data => data.suggestions)
                .catch(console.log)
    }

    getAddressSuggestions = async (inputValue) => {
        const dadataResponse = await this.sendDadataRequest(inputValue, 5);
        return dadataResponse.map(elem => elem.value);
    }

    renderSuggestions = async ({ value }) => {
        if (this.state.searchTerm !== this.state.inputValue) {
            value = this.state.searchTerm;
        };
        this.setState({ searchSuggestions: await this.getAddressSuggestions(value) });
    }

    clearSuggestions = () => {
        this.setState({ searchSuggestions: [] });
    }

    handleSearchSubmit = async () => {
        let dadataResponse = await this.sendDadataRequest(this.state.searchTerm, 1);
        if (this.state.startState) this.setState({ startState: false });
        if (dadataResponse[0]) {
            const location = dadataResponse[0];
            const { geo_lat, geo_lon } = location.data;
            const fiasLevel = +location.data.fias_level;
            if (fiasLevel < 8) {
                if (fiasLevel === 7) this.setState({ mapCoords: [geo_lat, geo_lon] });
                this.setState({ addressNotFound: true });
            } else {
                this.setState({ 
                    mapCoords: [geo_lat, geo_lon],
                    addressNotFound: false,
                    mapPinAddress: location.unrestricted_value,
                    infoBlockAddress: `${location.data.city}, ${location.value}`
                });
            };
        } else {
            this.setState({ addressNotFound: true });
        }
    }

    saveInitialInput = (event) => {
        const { searchSuggestions, inputValue } = this.state;
        if (searchSuggestions && !searchSuggestions.includes(inputValue)) {
            this.setState({ savedInput: event.target.value });
        }
    }

    fillInitialInput = () => {
        const { savedInput } = this.state;
        if (savedInput) {
            this.setState({ 
                            searchTerm: savedInput, 
                            inputValue: savedInput,
                            savedInput: "" 
            });
        }
    }

    handleCityChoiceOnMap = (event) => {
        const selectedCity = event.get("target").data._data;
        const { nameRus, country } = cityList.filter(city => city.location === selectedCity.center)[0];
        this.setState({ 
                        mapCoords: selectedCity.center,
                        mapPinAddress: `${nameRus}, ${country}`,
                        locationInEnglish: [selectedCity.content, ""],
                        locationInRussian: [nameRus, country]
                    });
    }

    render() {
        return (
            <YMaps query={{ apikey: this.yandexAPIKey, lang: 'en_RU' }}>
                <div className="explore-main-div">
                    <Helmet>
                        <title>Explore | House Explorer</title>
                    </Helmet>
                    <div className="explore-content">
                        <SearchBar 
                            handleInput={this.handleSearchInput}
                            handleSubmit={this.handleSearchSubmit}
                            inputValue={this.state.inputValue}
                            clearInput={this.clearSearchTerm}
                            searchSuggestions={this.state.searchSuggestions}
                            renderSuggestions={this.renderSuggestions}
                            clearSuggestions={this.clearSuggestions}
                            saveInitialInput={this.saveInitialInput}
                            fillInitialInput={this.fillInitialInput}
                        />
                        <HouseInfoBlock 
                            address={this.state.infoBlockAddress}
                            startState={this.state.startState}
                            addressNotFound={this.state.addressNotFound}
                        />
                    </div>
                    <React.Suspense fallback={<div></div>}>
                            <YandexMap 
                                mapCoords={this.state.mapCoords}
                                mapAddress={this.state.mapPinAddress}
                                startState={this.state.startState}
                                cityList={cityList}
                                locationInEnglish={this.state.locationInEnglish}
                                handleCityChoice={this.handleCityChoiceOnMap}
                            />
                    </React.Suspense>
                </div>
            </YMaps>
        );
    }
}

export default ExplorePage;