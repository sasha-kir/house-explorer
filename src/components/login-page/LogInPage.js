import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Helmet } from "react-helmet";
import './LogInPage.css';

class LogInPage extends Component {

    hidePlaceholder = (event) => {
        event.target.placeholder = "";
    }

    showPlaceholder = (event, placeholder) => {
        event.target.placeholder = placeholder;
    }

    render() {
        return (
            <div className="login-main-div">
                <Helmet>
                    <title>Log In | House Explorer</title>
                </Helmet>
                <div className="login-wrapper">
                    <div className="login-form">
                        <h2 className="form-header">log in</h2> 
                        <input className="form-element" 
                               onFocus={this.hidePlaceholder} 
                               onBlur={(e) => this.showPlaceholder(e, "username")}
                               placeholder="username" type="text" />
                        <input className="form-element" 
                               onFocus={this.hidePlaceholder}
                               onBlur={(e) => this.showPlaceholder(e, "password")}
                               placeholder="password" type="password" />
                        <button className="form-button">submit</button>
                        <p className="link-to-other-form">
                            Don't have an account? &nbsp;
                            <Link to="/register">Sign up</Link>
                        </p>
                    </div>
                </div>
            </div>
        );
    }
}

export default LogInPage;