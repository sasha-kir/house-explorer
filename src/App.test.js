import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { act } from 'react-dom/test-utils';

import App from './App';
import Navigation from './components/navigation/Navigation';
import PublicHomePage from './components/public-home-page/PublicHomePage';
import PlaceholderPage from './components/placeholder-page/PlaceholderPage';
import LogInPage from './components/login-page/LogInPage';
import RegisterPage from './components/register-page/RegisterPage';
import ExplorePage from './components/explore-page/ExplorePage';
import NotFoundPage from './components/404-page/NotFoundPage';

describe ('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <MemoryRouter>
        <App />
      </MemoryRouter>, div);
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders 404 page for invalid path', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/random' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(PublicHomePage)).toHaveLength(0);
    expect(wrapper.find(NotFoundPage)).toHaveLength(1);
  });
  it('renders public home page for default path', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(PublicHomePage)).toHaveLength(1);
  });
  it('renders placeholder page for about & cities paths', () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={[ '/about' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(PublicHomePage)).toHaveLength(0);
    expect(wrapper.find(PlaceholderPage)).toHaveLength(1);
    wrapper = mount(
      <MemoryRouter initialEntries={[ '/cities' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(PublicHomePage)).toHaveLength(0);
    expect(wrapper.find(PlaceholderPage)).toHaveLength(1);
  });
  it('renders login page', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/login' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(LogInPage)).toHaveLength(1);
  });
  it('renders register page', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[ '/register' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(RegisterPage)).toHaveLength(1);
  });
  it('renders explore page only when logged in, else renders login page', () => {
    let wrapper = mount(
      <MemoryRouter initialEntries={[ '/explore' ]}>
        <App/>
      </MemoryRouter>
    );
    expect(wrapper.find(ExplorePage)).toHaveLength(0);
    expect(wrapper.find(LogInPage)).toHaveLength(1);
  });
})
