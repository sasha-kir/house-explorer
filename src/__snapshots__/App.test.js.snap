// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <p>
      Hello World!
    </p>
  </header>
</div>
`;
